aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -1096,6 +1096,119 @@ Tensor clone(const Tensor& src, c10::optional<c10::MemoryFormat> optional_memory
   return self;
 }
 
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ covariations ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor cov(const Tensor& self, 
+           const c10::optional<Tensor>& otherOpt,
+           c10::optional<bool> rowvarOpt, 
+           c10::optional<bool> biasOpt, 
+           c10::optional<int64_t> ddofOpt,
+           const c10::optional<Tensor>& fweightsOpt, 
+           const c10::optional<Tensor>& aweightsOpt) {
+  
+  const Tensor other = otherOpt.value_or(Tensor());
+  const Tensor fweights = fweightsOpt.value_or(Tensor());
+  const Tensor aweights = aweightsOpt.value_or(Tensor());
+  auto bias = biasOpt.value_or(false);
+  auto ddof = ddofOpt.value_or(0);
+  auto rowvar = rowvarOpt.value_or(true);
+
+  if(self.size(0) == 0)
+    return at::empty_like(self);  
+  TORCH_CHECK(self.dim() <= 2,
+              "Input1 has more than 2 dimensions." );
+    
+  Tensor X, y;
+  if(self.dim() == 1)
+    X = self.unsqueeze(0);
+  else
+    X = self;
+
+  if(!rowvar && X.size(0) != 1)
+    X = X.t();
+
+  if(other.defined()) {
+    TORCH_CHECK(other.dim() <= 2,
+                "Input2 has more than 2 dimensions." );
+ 
+    if(other.dim() == 1)
+      y = other.unsqueeze(0);
+    else
+      y = other;
+
+    if(!rowvar && y.size(0) != 1)
+      y = y.t();
+
+    X = at::cat({X, y}, 0);
+  }
+
+  if (ddofOpt == c10::nullopt)
+    ddof = (bias == false) ? 1 : 0;
+  
+  Tensor w = at::empty(0, self.options());  
+  if(fweights.defined()) {
+    TORCH_CHECK((fweights.scalar_type() == at::kLong || fweights.scalar_type() == at::kInt),
+                "fweights must be integer.");
+    TORCH_CHECK(fweights.dim() <= 1,
+                "cannot handle multidimensional fweights.");
+    TORCH_CHECK(fweights.size(0) == X.size(1),
+                "incompatible numbers of samples and weights.");
+
+    TORCH_CHECK(fweights.min().item<float>() >= 0, 
+                "fweights cannot be negative.");
+    w = fweights;
+  }
+
+  if (aweights.defined()) {
```

Comment:
I'm going top stop reading through the implementation here for the moment. Take a look at the previous comments. cov() is a tricky function with a lot of inputs, but I think this can be simplified. Breaking it into conceptual blocks with comments ahead of each block might be helpful. For example, maybe the function can start by validating all its inputs. Currently the validation is scattered throughout. Then the initial values can be populated. Then the cov computation can occur.

Reply:
@mruberry 
That sounds like a great plan, I shall incorporate the block and comments. However, since few of the args such as `fweights` and `aweights` or intermediate vars such as `w_sum` can only be validated after performing some computations or changes to initial variables, they seem to be scattered everywhere. Owing to this dependency I dont think those can be moved to the top. However breaking it down to blocks seems doable!

Reply:
Makes sense. I look forward to seeing the update!

Reply:
@mruberry PTAL, do let me know if I should add in more details in the comments.